### ISSUE TITLE
Ansible include vs. include_tasks with tags

### ISSUE THREAD

**Issue by sapman** [ansible/ansible#32015]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

include_tasks
include
import 
import_tasks
tags

##### ANSIBLE VERSION

```
ansible 2.4.0.0
  config file = /opt/demansi/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### OS / ENVIRONMENT

Ubuntu 16.04

##### SUMMARY

I used to use include to include .yml files. I'm using tags all over my playbooks and roles. The only task I didn't tag are includes, cause they just apply their tag to all their tasks and they are included anyway.
When I upgraded to ansible 2.4, I started getting those deprecation messages that I should use include_tasks or import_tasks instead of include. 
I read the docs : https://docs.ansible.com/ansible/2.4/playbooks_reuse.html#differences-between-static-and-dynamic, and I saw `For dynamic includes, the task options will only apply to the dynamic task as it is evaluated, and will not be copied to child tasks.`
So I switched to include_tags, and when I put the always tag on the include_tasks, all the child tasks are executed with the always tag (Which is against the docs as I see it), and when I don't put tag to the include_tasks task, none of the child tasks are executed. 

##### STEPS TO REPRODUCE

play.yml
```yaml
---
- hosts: all
  gather_facts: no
  tasks:

  - name: "Include tasks file"
    include: tasks.yml

  - name: "Include_tasks tasks file"
    include_tasks: tasks.yml

```
tasks.yml
```yaml
- name: "Debug tag1"
  debug:
    msg: "init"
  tags:
  - tag1
```
##### EXPECTED RESULTS

I expect that include and include_tasks will behave the same way, and when running `ansible-playbook play.yml -t tag1`, the debug will happen twice.

##### ACTUAL RESULTS

Using include_tasks give different results than include.
```
$ansible-playbook play.yml -t tag1 -vvvv

ansible-playbook 2.4.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
Using /etc/ansible/ansible.cfg as config file
setting up inventory plugins
Set default localhost to localhost
Parsed /etc/ansible/hosts inventory source with ini plugin
[DEPRECATION WARNING]: The use of 'include' for tasks has been deprecated. Use 'import_tasks' for static inclusions or 'include_tasks' for dynamic inclusions. This feature will be removed in a future 
release. Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
[DEPRECATION WARNING]: include is kept for backwards compatibility but usage is discouraged. The module documentation details page may explain more about this rationale.. This feature will be removed in 
a future release. Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
statically imported: /home/user/tasks.yml
Loading callback plugin default of type stdout, v2.0 from /usr/lib/python2.7/dist-packages/ansible/plugins/callback/__init__.pyc

PLAYBOOK: play.yml *****************************************************************************************************************************************************************************************
1 plays in play.yml

PLAY [all] *************************************************************************************************************************************************************************************************
META: ran handlers

TASK [Debug tag1] ******************************************************************************************************************************************************************************************
task path: /home/user/tasks.yml:2
ok: [localhost] => {
    "msg": "init"
}
META: ran handlers
META: ran handlers

PLAY RECAP *************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0   

```

The only workaround I found for now is running `ansible-playbook play.yml -t untagged,tag1`, this will import the files, but won't apply its tags on them (Assuming the only untagged tasks are include_tasks).

**Comment by jurgenhaas** [comment]
I do have the same problem and it's really hitting me hard.

And I guess there is an additional and cloely related bug too. I thought a workaround could be to give all my `include_tasks` the tag `always` as the [documentation](https://docs.ansible.com/ansible/2.4/playbooks_reuse.html#differences-between-static-and-dynamic) states this:

```
When it comes to Ansible task options like tags and conditional statements (when:):

For static imports, the parent task options will be copied to all child tasks contained within the import.
For dynamic includes, the task options will only apply to the dynamic task as it is evaluated, and will not be copied to child tasks.
```

But in reality, include_tasks DOES copy all it's tags down to all children.

**Comment by ChiKenNeg** [comment]
Same here, it's kinda hard to maintain a fully tagged ansible code without the basic functionality to control desired 'include' flows.

**Comment by Chrislevi** [comment]
Very confusing, I don't find putting "always" in all include_tasks as a valid workaround. 
considering to change my whole automation to a different tool despite of ansible being very appealing and flawless in certain areas.

**Comment by chrnie** [comment]
`ansible-playbook play.yml -t untagged,tag1`  is a quite good workaround, but this should be the default behaviour in my opinion.

**Comment by jurgenhaas** [comment]
Don't think this is a valid workaround. Because I have tons of untagged tasks that should not be executed, only the tagged ones. What used to be executed while it was untagged, was the `include` statement and that's now broken badly.

I don't understand why the response from the developers is so slow on this and the tags on this issue are worrysome as well. Don't see this being a support issue. It's a badly broken feature which breaks almost all of my plays.

**Comment by bcoca** [comment]
So #32769 should fix the issue with tags in dynamic tasks, in any case the original issue should be using `import_tasks` instead which works w/o having to add a tag to that task:

```
---
- hosts: all
  gather_facts: no
  tasks:

  - name: "Include tasks file"
    include: tasks.yml

  - name: "import_tasks tasks file"
    import_tasks: tasks.yml
```

If you need the include to be dynamic, the PR above will allow you to add tags to the task w/o the included tasks inheriting them:
```
  - name: "import_tasks tasks file"
    include_tasks: tasks.yml
    tags:
       - tag1
 
```

tasks.yml
```
- debug: msg='i wont run when tag1 is used'
- debug: msg='I run when tag1 is used'
  tags:
    - tag1
```

**Comment by jurgenhaas** [comment]
Thanks @bcoca I can confirm that the case with `import_tasks` is now working as expected in version 2.4.2.0 - the original report was against 2.4.0.0 so it looks like that got fixed in the meantime.

What's not working as expected yet is the dynamic `include_task`. In your example the tag is inherited to all tasks in the included file which is not what you'd expect. Let's have a look at this example:

```
- hosts: all
  gather_facts: no
  tasks:
  - include_tasks: tasks2.yml
    with_items:
      - 'abc'
      - 'def'
```

With tasks2.yml looking like this:

```
- debug: msg="Task 2 without tag {{ item }}"

- debug: msg="Task 2 {{ item }}"
  tags:
    - tag1
```

When calling this playbook without tags, both tasks get executed twice. When using `--tags=tag1` then nothing gets called.

Is that case not fixed yet or just not included in 2.4.2.0 yet?

**Comment by jurgenhaas** [comment]
OK, looks like this is not yet part of 2.4.2.0. So I have patched my Ansible with #32769 and changed my playbook to this:

```
- hosts: all
  gather_facts: no
  tasks:
  - include_tasks: tasks2.yml
    with_items:
      - 'abc'
      - 'def'
    tags:
      - tag1
```

But that still executes all tasks in the included files. It doesn't seem to turn inheritance off.

**Comment by samdoran** [comment]
`include` and `include_tasks` are not equivalent. `include` is equivalent to `import_tasks` since they are both _static_. 

`include_tasks` is dynamic. Since there are no tags applied to to `include_tasks` task, the inclusion of the file is skipped, and the tags on tasks inside that file are not available.

We split these out hoping to make the behavior more obvious and consistent, but it's still a bit confusing.

To leverage tags on tasks inside of task files, they will have to be statically imported. Changing `include` to `import_tasks` will let you use tags inside statically imported files.

All that being said, something does look wrong with `include_tasks`. Tags added to the include should not be applied to tasks that are dynamically included. Any tag I set on `include_tasks` seems to cause all tasks within the included file to run. That seems incorrect.

**Comment by samdoran** [comment]
Sorry for the noise. That's what I get for not refreshing my tabs before commenting. :blush:

**Comment by oukooveu** [comment]
Original issue was not resolved and run of `ansible-playbook play.yml -t tag1` executes debug task only once. Please re-open.
